project/src/main/java/com/google/sps/data/ServletLibrary.java

```
@@ -276,8 +396,49 @@ public static void addProductToProductCategory(DatastoreService datastore, Strin
     }
   }
 
+  // Delete a product from the specified product category.
+  public static void deleteProductFromProductCategory(DatastoreService datastore, 
+                                                      String productId, 
+                                                      String productCategory) {
+    if (datastore == null || productId == null || productCategory == null) {
+      System.err.println("At least one of the inputs was null!");
+      return;
+    }                                                  
+    Filter filter = new FilterPredicate("productCategory", FilterOperator.EQUAL, productCategory);
+    Query query = new Query("ProductCategory").setFilter(filter);
+    PreparedQuery pq = datastore.prepare(query);
+    Entity entity = pq.asSingleEntity();
+    // If the category exists, remove the product.
+    if (entity != null) {
+      @SuppressWarnings("unchecked") // Documentation says to suppress warning this way
+        List<String> productIds = (ArrayList<String>) entity.getProperty("productIds"); 
+      if (productIds == null) productIds = new ArrayList<String>();
+      productIds.remove(productId);
+      entity.setProperty("productIds", productIds);
+      datastore.put(entity);
+    }
+  }
+
+  public static void updateProductCategories(DatastoreService datastore, 
+                                       String productId, 
+                                       String oldProductCategory,
+                                       String productCategory) {
+    if (datastore == null || productId == null || productCategory == null || 
+        oldProductCategory == null) {
+      System.err.println("At least one of the inputs was null!");
+      return;
+    }
```

Might not be the best way to handle this since it will fail silently and the user won't know their update has failed.  It might be better to consider throwing a RuntimeException or InvalidArgumentException or alternatively returning true/false if the request succeeded.

---

Good point, I'll change it to return an exception.